### ISSUE TITLE
[no-unnecessary-condition] False positive when using `while` loop

### ISSUE THREAD

**Issue by sqren** [typescript-eslint/typescript-eslint#2262]
I'm running into an issue where `@typescript-eslint/no-unnecessary-condition` is giving a false positive:
![image](https://user-images.githubusercontent.com/209966/85932967-d043c780-b8d1-11ea-9943-affbbf917ff9.png)

**Repro**
```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

```TS
async function helloWorld() {
  let myFlag = false;
  setTimeout(() => {
    myFlag = true;
  }, 100);

  while (myFlag === false) {
    await new Promise((resolve) => setImmediate(resolve));
  }
}
```

**Expected Result**
I shouldn't see any eslint errors

**Actual Result**
I see the error: 
```
Unnecessary conditional, both sides of the expression are literal values
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.4.0` |
| `@typescript-eslint/parser`        | `3.4.0` |
| `TypeScript`                       | `3.9.5` |
| `ESLint`                           | `7.3.1` |
| `node`                             | `10.21.0` |
| `npm`                              | `6.14.4` |

**Comment by bradzacher** [comment]
See discussion #1459

As per your screenshot - TS reports the tpye of `myFlag` to be `false`.
When the rule interrogates the type, it gets told it's `false`, so the rule sees the condition as `false === false` which is rightfully flagged.
